fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/RecordCursorTest.java

```diff
@@ -1277,28 +1277,46 @@ void asStreamTest() {
 
     }
 
-    @Test
-    public void mapPipelineCloseWhileCancelling() {
+    @Nonnull
+    private RecordCursor<Integer> mapPipelinedCursorToClose(int iteration, CompletableFuture<Void> signal) {
+        return RecordCursor.fromList(EXECUTOR, IntStream.range(0, iteration % 199).boxed().collect(Collectors.toList()))
+                .mapPipelined(val -> signal.thenApplyAsync(ignore -> val + 349, EXECUTOR), iteration % 19 + 2);
+    }
+
+    @Nonnull
+    private RecordCursor<String> flatMapPipelinedCursorToClose(int iteration, CompletableFuture<Void> signal) {
+        return RecordCursor.flatMapPipelined(
+                outerContinuation -> RecordCursor.fromList(EXECUTOR, IntStream.range(0, iteration % 199).boxed().collect(Collectors.toList()), outerContinuation),
+                (outerValue, innerContinuation) -> new FutureCursor<>(EXECUTOR, signal.thenApplyAsync(ignore -> String.valueOf(outerValue), EXECUTOR)),
+                null,
+                null,
+                iteration % 19 + 2
+        );
+    }
+
+    private <T> void closePipelineWhileCancelling(@Nonnull BiFunction<Integer, CompletableFuture<Void>, RecordCursor<T>> cursorGenerator) {
         Map<Class<? extends Throwable>, Integer> exceptionCount = new HashMap<>();
         for (int i = 0; i < 20_000; i++) {
             try {
-                LOGGER.info(KeyValueLogMessage.of("running map pipeline close test", "iteration", i));
+                LOGGER.info(KeyValueLogMessage.of("running pipeline close test", "iteration", i));
                 CompletableFuture<Void> signal = new CompletableFuture<>();
-                RecordCursor<Integer> cursor = RecordCursor.fromList(IntStream.range(0, i % 199).boxed().collect(Collectors.toList()))
-                        .mapPipelined(val -> signal.thenApplyAsync(ignore -> val + 349, EXECUTOR), i % 19 + 2);
-                CompletableFuture<RecordCursorResult<Integer>> resultFuture = cursor.onNext();
+                RecordCursor<T> cursor = cursorGenerator.apply(i, signal);
+                CompletableFuture<RecordCursorResult<T>> resultFuture = cursor.onNext();
 
                 signal.complete(null);
                 cursor.close();
                 resultFuture.get(2, TimeUnit.SECONDS);
             } catch (Exception e) {
                 Throwable errToCount;
-                if (e instanceof CompletionException && e.getCause() != null) {
+                if (e instanceof ExecutionException && e.getCause() != null) {
                     errToCount = e.getCause();
                     errToCount.addSuppressed(e);
                 } else {
                     errToCount = e;
                 }
+                if (errToCount instanceof CancellationException) {
+                    continue;
+                }
```

Comment:
`MapPipelinedCursor`, I think, never throws `CancellationException` in this test.
Did you see that?
The code has paths where it could, so if it is not, that's probably a shortcoming of the test.